activesupport/lib/active_support/core_ext/enumerable.rb

```diff
@@ -114,11 +114,15 @@ def sum(identity = nil)
 # just calling the compat method in the first place.
 if Array.instance_methods(false).include?(:sum) && !(%w[a].sum rescue false)
   class Array
-    remove_method :sum
+    alias :orig_sum :sum
 
-    def sum(*args) #:nodoc:
-      # Use Enumerable#sum instead.
-      super
+    def sum(init = nil, &block) #:nodoc:
+      if init.is_a?(Numeric) || self.first.is_a?(Numeric)
```

Comment:
I'm not sure which results do you want.

Rails master and Ruby 2.3 does not support `[1, "2"].sum`

```shell
$ git rev-parse --short HEAD
a4b3c78

$ cat array_sum_str.rb
require 'active_support/core_ext/enumerable'

ary = [1, "2"]
p ary.sum

$ bundle exec ruby -v -I lib array_sum_str.rb
ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-darwin14]
ruby/rails/activesupport/lib/active_support/core_ext/enumerable.rb:24:in `+': String can't be coerced into Fixnum (TypeError)
	from ruby/rails/activesupport/lib/active_support/core_ext/enumerable.rb:24:in `each'
	from ruby/rails/activesupport/lib/active_support/core_ext/enumerable.rb:24:in `inject'
	from ruby/rails/activesupport/lib/active_support/core_ext/enumerable.rb:24:in `sum'
	from array_sum_str.rb:4:in `<main>'
```